File: .github/workflows/script/rollback_changelog.py
import datetime
import os
import argparse

EMPTY_CHANGELOG = """# CodeQL Action Changelog

"""

def get_today_string():
  today = datetime.datetime.today()
  return '{:%d %b %Y}'.format(today)

# Include everything up to and after the first heading,
# but not the first heading and body.
def drop_unreleased_section(lines: list[str]):
  before_first_section = ''
  after_first_section = ''
  found_first_section = False
  skipped_first_section = False

  for i, line in enumerate(lines):
    if line.startswith('## ') and not found_first_section:
      found_first_section = True
    elif line.startswith('## ') and found_first_section:
      skipped_first_section = True

    if not found_first_section:
      before_first_section += line
    if skipped_first_section:
      after_first_section += line

  return (before_first_section, after_first_section)

def update_changelog(target_version, rollback_version, new_version):
  before_first_section = EMPTY_CHANGELOG
  after_first_section = ''

  if (os.path.exists('CHANGELOG.md')):
    with open('CHANGELOG.md', 'r') as f:
      (before_first_section, after_first_section) = drop_unreleased_section(f.readlines())

  newHeader = f'## {new_version} - {get_today_string()}\n'

  print(before_first_section, end="")
  print(newHeader)
  print(f"This release rolls back {rollback_version} due to issues with that release. It is identical to {target_version}.\n")
  print(after_first_section)

# We expect three version strings as input:
#
# - target_version: the version that we are re-releasing as `new_version`
# - rollback_version: the version that we are rolling back, typically the one that followed `target_version`
# - new_version: the new version that we are releasing `target_version` as, typically the one that follows `rollback_version`
#
# Example: python3 .github/workflows/script/rollback_changelog.py --target-version "1.2.3" --rollback-version "1.2.4" --new-version "1.2.5"
parser = argparse.ArgumentParser(description="Update CHANGELOG.md for a rollback release.")
parser.add_argument("--target-version", "-t", required=True, help="Version to re-release as new_version.")
parser.add_argument("--rollback-version", "-r", required=True, help="Version being rolled back.")
parser.add_argument("--new-version", "-n", required=True, help="New version to publish for target_version.")
args = parser.parse_args()

update_changelog(args.target_version, args.rollback_version, args.new_version)

File: .github/workflows/script/rollback_changelog.ts
#!/usr/bin/env ts-node

import * as fs from 'fs';

const EMPTY_CHANGELOG = `# CodeQL Action Changelog

`;

function getTodayString(): string {
  const today = new Date();
  return today.toLocaleDateString('en-GB', {
    day: '2-digit',
    month: 'short',
    year: 'numeric'
  });
}

// Include everything up to and after the first heading,
// but not the first heading and body.
function dropUnreleasedSection(lines: string[]): [string, string] {
  let beforeFirstSection = '';
  let afterFirstSection = '';
  let foundFirstSection = false;
  let skippedFirstSection = false;

  for (const line of lines) {
    if (line.startsWith('## ') && !foundFirstSection) {
      foundFirstSection = true;
    } else if (line.startsWith('## ') && foundFirstSection) {
      skippedFirstSection = true;
    }

    if (!foundFirstSection) {
      beforeFirstSection += line + '\n';
    }
    if (skippedFirstSection) {
      afterFirstSection += line + '\n';
    }
  }

  return [beforeFirstSection, afterFirstSection];
}

function updateChangelog(targetVersion: string, rollbackVersion: string, newVersion: string): void {
  let beforeFirstSection = EMPTY_CHANGELOG;
  let afterFirstSection = '';

  if (fs.existsSync('CHANGELOG.md')) {
    const content = fs.readFileSync('CHANGELOG.md', 'utf8');
    [beforeFirstSection, afterFirstSection] = dropUnreleasedSection(content.split('\n'));
  }

  const newHeader = `## ${newVersion} - ${getTodayString()}\n`;

  process.stdout.write(beforeFirstSection);
  console.log(newHeader);
  console.log(`This release rolls back ${rollbackVersion} due to issues with that release. It is identical to ${targetVersion}.\n`);
  process.stdout.write(afterFirstSection);
}

// Parse command line arguments
interface Args {
  targetVersion?: string;
  rollbackVersion?: string;
  newVersion?: string;
}

function parseArgs(): Args {
  const args: Args = {};
  
  for (let i = 2; i < process.argv.length; i++) {
    switch (process.argv[i]) {
      case '--target-version':
      case '-t':
        args.targetVersion = process.argv[++i];
        break;
      case '--rollback-version':
      case '-r':
        args.rollbackVersion = process.argv[++i];
        break;
      case '--new-version':
      case '-n':
        args.newVersion = process.argv[++i];
        break;
    }
  }
  
  return args;
}

// Main execution
const args = parseArgs();

if (!args.targetVersion || !args.rollbackVersion || !args.newVersion) {
  console.error('Usage: rollback_changelog.ts --target-version <version> --rollback-version <version> --new-version <version>');
  process.exit(1);
}

updateChangelog(args.targetVersion, args.rollbackVersion, args.newVersion);